base/errorshow.jl

```diff
@@ -208,6 +208,32 @@ function show_convert_error(io::IO, ex::MethodError, @nospecialize(arg_types_par
     end
 end
 
+"""
+    methoderror_hints
+
+Packages can provide hints about appropriate corrective actions for
+specific `MethodError`s with `push!(Base.methoderror_hints, hintmessage)`, where
+`hintmessage` should be a function like
+
+```julia
+function hintmessage(f, arg_types, kwargs)
+    # Test to see whether the call matches the specific pattern for this message
+    if f === myfunc && length(arg_types) == 1 && arg_types[1] <: SomeType
+        return "`myfunc(::SomeType)` is not defined, did you mean to call `otherfunc`?"
+    end
+    return nothing    # use `nothing` to indicate that f, arg_types, kwargs didn't match
+end
+```
+
+Packages should perform the `push!` onto `Base.methoderror_hints` from
+their `__init__` function.
+
+!!! compat "Julia 1.5"
+    Custom MethodError hints are available as of Julia 1.5.
+"""
+const methoderror_hints = []
+# Note: Base should not use `methoderror_hints`, it should inline
+# custom hints below to avoid the slight performance hit from `invokelatest`.
```

Comment:
Is the overhead from `invokelatest` significant here?

Reply:
Yes (in a relative sense, i.e., it's a large fraction of total cost), though acceptable IMO. I do think it's viable to have Base use `methoderror_hints` also, but since Base doesn't actually need to do so I thought it would be better to encourage continuing the current trajectory.

Another issue not yet discussed is where there are any circumstances where the callback needs to receive the argument values rather than just the types. The values are not always guaranteed to be available (see the comment re `invoke` at the top of `showerror`), whereas the types are.

Reply:
OK, in the new version I pass the exception object as well as some processed arguments, so I'm not so worried we'd want to change the API anymore.